#!/usr/bin/env bash

BWA_HOME=${HOME}/bwa
FREEBAYES_HOME=${HOME}/freebayes

if [[ $# -ne 2 ]]; then
    echo "Usage: run_variant_analysis.sh <hs38|hs38a|hs37> <sequence_prefix>"
    echo "       (example of a sequence prefix: SRR062635)"
    exit
fi

if [[ ! -f "${1}.fa" ]]; then
    echo "ðŸ˜¡ Missing reference genome. Run setup_reference_genome.sh."
    exit
fi

if [[ ! -f "${2}_1.filt.fastq.gz" || ! -f "${2}_2.filt.fastq.gz" ]]; then
    echo "ðŸ˜¡ Missing files for sequence ${2}. Download from ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase3/data/."
    exit
fi

echo "ðŸ‘‰ Starting alignment with bwa"
BWA_CMD="${BWA_HOME}/bwa mem ${1}.fa ${2}_1.filt.fastq.gz ${2}_2.filt.fastq.gz | gzip > ${2}.sam.gz"
eval ${BWA_CMD}
if [[ $? -eq 0 ]]; then
    echo "ðŸ‘‰ Done with alignment"
else
    echo "ðŸ˜¡ Failed running bwa"
    exit
fi

echo "ðŸ‘‰ Converting to BAM file"
SAM2BAM_CMD="samtools view -b ${2}.sam.gz > ${2}.bam"
eval ${SAM2BAM_CMD}
if [[ $? -eq 0 ]]; then
    echo "ðŸ‘‰ Done with BAM conversion"
else
    echo "ðŸ˜¡ Failed running BAM conversion"
    exit
fi

echo "ðŸ‘‰ Performing sorting of BAM file"
SORT_CMD="sambamba sort ${2}.bam"
eval ${SORT_CMD}
if [[ $? -eq 0 ]]; then
    echo "ðŸ‘‰ Done with sorting BAM file"
else
    echo "ðŸ˜¡ Failed sorting BAM file"
    exit
fi

echo "ðŸ‘‰ Marking duplicates in BAM file"
MARKDUP_CMD="sambamba markdup ${2}.sorted.bam ${2}.final.bam"
eval ${MARKDUP_CMD}
if [[ $? -eq 0 ]]; then
    echo "ðŸ‘‰ Done with marking duplicates in BAM file"
else
    echo "ðŸ˜¡ Failed marking duplicates in BAM file"
    exit
fi

echo "ðŸ‘‰ Running freebayes for variant calling"
FREEBAYES_CMD="${FREEBAYES_HOME}/bin/freebayes -f ${1}.fa ${2}.final.bam > ${2}.output.vcf"
eval ${FREEBAYES_CMD}
if [[ $? -eq 0 ]]; then
    echo "ðŸ‘‰ Done with variant calling. See ${2}.output.vcf file."
else
    echo "ðŸ˜¡ Failed performing variant calling"
    exit
fi

echo "ðŸ‘‰ Done!"